src/Servers.jl

```diff
@@ -130,278 +109,196 @@ function check_rate_limit(tcp;
+                reuse_limit::Int=nolimit, readtimeout::Int=0,
+                verbose::Bool=false)
 
-    if ssl && sslconfig === nothing
-        sslconfig = SSLConfig(require_ssl_verification)
-    end
+    handler = f isa Handler ? f : stream ? StreamHandlerFunction(f) : RequestHandlerFunction(f)
 
-    @info "Listening on: $host"
-    if isassigned(tcpref)
+    inet = getinet(host, port)
+    if tcpref !== nothing
         tcpserver = tcpref[]
     elseif reuseaddr
         tcpserver = Sockets.TCPServer(; delay=false)
-        if Sys.islinux() || Sys.isapple()
+        if Sys.isunix()
             rc = ccall(:jl_tcp_reuseport, Int32, (Ptr{Cvoid},), tcpserver.handle)
-            Sockets.bind(tcpserver, host.host, host.port; reuseaddr=true)
+            Sockets.bind(tcpserver, inet.host, inet.port; reuseaddr=true)
         else
             @warn "reuseaddr=true may not be supported on this platform: $(Sys.KERNEL)"
-            Sockets.bind(tcpserver, host.host, host.port; reuseaddr=true)
+            Sockets.bind(tcpserver, inet.host, inet.port; reuseaddr=true)
         end
         Sockets.listen(tcpserver)
     else
-        tcpserver = Sockets.listen(host)
-        tcpref[] = tcpserver
+        tcpserver = Sockets.listen(inet)
     end
+    verbose && @info "Listening on: $host:$port"
 
-    if host isa Sockets.InetAddr # build debugging info
-        hostname = string(host.host)
-        hostport = string(host.port)
-    else
-        hostname = string(host)
-        hostport = ""
+    if tcpisvalid === nothing
+        tcpisvalid = ratelimit === nothing ? x->true : x->check_rate_limit(x, ratelimit)
     end
-    listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
+
+    return listenloop(handler, Server2(sslconfig, tcpserver, string(host), string(port)), tcpisvalid,
+        connectioncounter, reuse_limit, readtimeout, verbose)
 end
 
-function listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
-    try
-        id = 0
-        while isopen(tcpserver)
-            try
-                io = accept(tcpserver)
-                if !tcpisvalid(io; kw...)
-                    @info "Accept-Reject:  $io"
-                    close(io)
-                    continue
-                end
-                io = getsslcontext(io, sslconfig)
-                let i=id, conn = Connection(hostname, hostport, pipeline_limit, 0, require_ssl_verification, io)
-                    @async try
-                        @info "Accept ($i):  $conn"
-                        connectioncounter[] += 1
-                        handle_connection(f, conn; kw...)
+"main server loop that accepts new tcp connections and spawns async threads to handle them"
+function listenloop(h, server,
+    tcpisvalid=x->true, connectioncounter=Ref(0),
+    reuse_limit::Int=1, readtimeout::Int=0, verbose::Bool=false)
```

Comment:
Why the defaults here ? Is there any intention that listenloop would ever be called anywhere other than just above? I though this was just a work-around for a compiler limitation.

The `Server2` type seems a little inelegant. Is this needed to coerce the compiler into doing what we want?

Is there a more elegant solution? Could we pass the big-collection-of-optional-stuff through the stack as a named-tuple?

Is there still a performance problem with passing around kw args and only declaring the ones used in each function? I have some memory that kw args were supposed to become 1st class in Julia 1.0.